_posts/2016-07-07-better-email-validation-with-obey-and-mailgun.markdown

```diff
@@ -0,0 +1,68 @@
+
+Basically, with all the special characters and conditions of what can go before and after the `@`, a regex to validate emails accurately would be incomprehensable at best.
+
+_If you'd like to find out more about why email validation sucks check out [this article](http://www.kbedell.com/2011/03/16/how-to-validate-an-email-address-using-regular-expressions/), for now let's move along._
+
+Surely this is something other people have dealt with though, what about a library? Well, they [do exist](https://www.npmjs.com/package/isemail), however, we want to _really_ validate an email. This is most likely either a core communication or authentication property. Libraries can do more than a regex, but they lack more advanced checks against actual resources. Testing for things like not only the validity of a domain, but does it even exist.
+
+### Using a Service for Validation
+
+We want to know if an email is valid, but a regex's are tricky and libraries are limited. This is where validation services come in handy. [Mailgun's API for email validation](https://documentation.mailgun.com/api-email-validation.html#email-validation) is a wonderful option. It will check a number of things for us:
+
+* General syntax based on RFC
+* DNS validation
+* Spell check (common email services)
+* Service Provider checking
+
+Not only will it take care of the general syntax (what we _would_ use a hyper-complex regex for) but it will make "best-attempt" tries at checking for common errors and valid domains.
+
+So, you may be asking yourself what this looks like. Typically validation is a synchronous task but we're going to call an API to make this work. This is where [Obey](https://github.com/TechnologyAdvice/obey) makes things simple. The promise-first design allows it to handle async and sync requests exactly the same. But what does this look like?
+
+We start with a simple type definition, using Obey and [request-promise](https://www.npmjs.com/package/request-promise) we'll hit the API and handle the results:
+
+```javascript
+const obey = require('obey')
+const rp = require('request-promise')
+
+// Our API key
+const mailgunPublicKey = 'XXXXX-YYYY-ZZZZZZZ'
+
+obey.type('mailgunEmail', (context) => {
+    // Hit the API
+    return rp(`https://api:${mailgunPublicKey}'@api.mailgun.net/v3/address/validate?address=${encodeURIComponent(context.value)}`)
+        .then((result) => {
+            // Handle validation errors
+            if (!JSON.parse(result).is_valid) {
+                context.fail(context.value + ' is an invalid email')
+            }
+        })
+})
+```
+
+That's it. Now when we define a model or rule we can use the `mailgunEmail` type and validation against the Mailgun API will either pass (no action) or fail (call to `context.fail` resulting in a `ValidationError` throw).
```

Comment:
Can't figure out how to read this.  I think it's two sentences or more?

>Now when we define a model or rule we can use the `mailgunEmail` type and validation against the Mailgun API will either pass (no action) or fail (call to `context.fail` resulting in a `ValidationError` throw).


Might be:

>Now when we define a model **or rule, we can** use the `mailgunEmail` type and validation against the **Mailgun API.  It will** either pass (no action) or fail (call to `context.fail` resulting in a `ValidationError` throw).